### DISCUSSION TITLE
Bug? Xdrip sends the 5 minute backfilled BG to AAPS which causes issues

### DISCUSSION THREAD

**Discussion by nlof** [NightscoutFoundation/xDrip#2267] [General]
If xdrip misses a BG due to BT connection issues (Dexcom G5), next time it connects, it backfills BG.
AAPS gets the current BG for a split second, after which it receives the backfilled BG from Xdrip and stays with that 5 minute old BG until the next reading.
I've seen that frequently because my new phone has frequent xdrip BT connection issues with my G5 transmitter, basically every 10-15 minutes... (The bad BT connectivity on the Xdrip side is a separate issue).

See screenshots of 2x AAPS sequentially and Xdrip after such an event. It last occurred @ 19:31 in the attached logs on AAPS bug report: https://github.com/nightscout/AndroidAPS/issues/2004 . 
I see that I have to email the Xdrip logs? 
I'm using Xdrip [nightly built Aug 3rd 2022](https://github.com/NightscoutFoundation/xDrip/releases/tag/2022.08.03), AAPS 3.1.0.3 and Dexcom G5.

**Comment by Navid200** [comment]
>  (The bad BT connectivity on the Xdrip side is a separate issue).

What do you mean by that?
Does your phone have a connectivity issue with the transmitter or not?
If it does and if you can resolve it by correctly setting up your phone, why is that not a solution?

**Reply by nlof** [reply]
It is not a solution it just lessens the impact of this bug. Connectivity issue is happening much more frequently with the new phone but it used to happen on occasion with the old one as well. I will try to solve it but it still doesn't solve the basic problem of Xdrip sending the backfilled BG to AAPS when this happens.

**Reply by Navid200** [reply]
You say AAPS receives a backfill from xDrip and stays with it for 5 minutes.  What is it supposed to do?
Are you saying when the time comes for the next reading in 5 minutes, you don't want AAPS to accept it?  Why?

**Reply by nlof** [reply]
AAPS retains the backfill INSTEAD of the new reading. There's a new reading on Xdrip which gets to AAPS but it gets wiped out immediately by the backfilled BG also read at the same time by Xdrip. So AAPS retains or uses the 5 minute old BG despite the fact that Xdrip has just read both the 5 minute old and the new current BG. 
People on AAPS are saying that Xdrip shouldn't send the backfill to AAPS at all. Just the new reading.
Or maybe if the sequence matters, maybe send the backfill first and then a fraction of a second later send the new reading so that the last reading that remains is the new one.

**Reply by Navid200** [reply]
Is the following what's happening?

AAPS receives a reading from xDrip with a time parameter of 8:00 AM.
Shortly after, AAPS receives a backfill reading with a time parameter of 7:55.
AAPS deletes the reading for 8:00 and only shows the reading for 7:55?

**Reply by nlof** [reply]
Yes, exactly.

**Reply by nlof** [reply]
Not sure AAPS deletes the 8:00 from its graphic, but it displays the 7:55 digit as the latest value and the 7:55 timestamp while xdrip shows the new 8:00 value and timestamp. When looking carefully at AAPS, one can see the 8:00 value first which gets wiped out a few seconds later by the 7:55 value.

**Reply by Navid200** [reply]
Don't you think that's an AAPS bug?

**Reply by nlof** [reply]
Good question. AAPS people pointed to Xdrip and vice versa :)
I'm not sure but it's some kind of a bug between the two apps :)
Maybe between the two it can be solved somehow :)

**Reply by Navid200** [reply]
It's definitely an xDrip bug if xDrip reports an incorrect time parameter.

I have been using G6 for about 4 years now using xDrip and I have never seen a backfill overwrite a different reading.

If someone is telling you that this is an xDrip bug, please ask them to post in this thread and give us more detail.

**Reply by justmara** [reply]
For me it looks more like AAPS bug - it processes all the data received from xDrip, even backfilled one, and sends all the values as events for Overview screen refresh. And all the events are processed there to redraw main screen. So since backfilled values comes later (even with correct timestamp) - main screen (overview) refreshes to reflect this (latest to come) value, event if it already received newer data and have already redrawn with it. 
This looks exactly like the behaviour described: new data comes, but then switches to older until newer data comes.
Simple `does not refresh, if this event is older than one already used to redraw` would totally fix this behavior,  imo.
Am I wrong?

**Reply by robertrub** [reply]
Navid, maybe my screen captures can explain. The xDrip data arrives to AAPS, is sorted correctly in the xDrip tab. But, it is (maybe) the first of the BG series that arrived that is used by AAPS for the calculationd
![SAVE_20220815_214749](https://user-images.githubusercontent.com/76626474/184707059-6580f7e9-74cb-419f-ab89-e2089dae21aa.jpg)
![SAVE_20220815_214819](https://user-images.githubusercontent.com/76626474/184707061-00989049-f6ab-4b2d-820a-112096a7e703.jpg)
![SAVE_20220815_214845](https://user-images.githubusercontent.com/76626474/184707062-7869c23c-a319-453b-bbdd-11af6c632ccb.jpg)
![SAVE_20220815_214922](https://user-images.githubusercontent.com/76626474/184707065-235d31ab-c789-4f33-9584-9daa8280e1d4.jpg)

**Reply by Navid200** [reply]
Does the same not happen if you have backfill while using Dexcom app?

**Reply by robertrub** [reply]
Yes, Dexcom fills in the missed BGs too. No, AAPS takes the last BG (time stamp wise) to make the calculations.

**Reply by gshry** [reply]
hi, One plus 9 pro - same problem , (AAPS 3.1.0.3)

**Reply by andyrozman** [reply]
@gshry Try to use latest nightly and enable Samsung workaround. This will make Xdrip work better on OP, and it might also solve this problem (see discussion about that here)... When I was using stable(er) version (withoit fix), I was not just missing reading I was also getting errors in AAPS (yellow triange - recalculated value used)

**Reply by gshry** [reply]
[Nightly build 20th Feb 2023 & enable Samsung workaround, didn't solve the problem.

**Reply by andyrozman** [reply]
Are you using all the recommended (default) settings? Some people with Oxygen 12.x or higher (I am on 13) who tried new version and Samsung workaround are reporting that everything started working for them.

**Comment by Navid200** [comment]
I have no reason to believe there has been a change.
But, would you please try the latest Nightly, released today, and let us know?

**Reply by robertrub** [reply]
Tried with nightly build of the19th with the same result

![Screenshot_2022-08-19-16-16-32-569_info nightscout androidaps](https://user-images.githubusercontent.com/76626474/185639058-d85c0104-9c9c-48bd-9989-71aa03a72212.png)

![IMG_20220819_161705](https://user-images.githubusercontent.com/76626474/185639106-c526a0d9-9012-4d9d-8560-cf4c8c017a4a.jpg)

**Comment by justmara** [comment]
this [commit](https://github.com/nightscout/AndroidAPS/commit/8fb5c5a984a8e16062e70282ce3d0fd6383ed864#diff-cbe72af028b300d0003dec22f8eb1f1201fca15a09c2bb4c0d3b1ea2eb344ff8R19) in AAPS dev seems to be related :)

**Reply by asquelt** [reply]
definitely an AAPS fault to assume time always runs forward :)

**Reply by andyrozman** [reply]
Xdrip should not send "backsies" in broadcast, it should only save them in NS.
